webapp/sources/rudder/rudder-core/src/test/scala/com/normation/rudder/services/policies/TestWriteNodeCertificatesPem.scala

```diff
@@ -0,0 +1,193 @@
+              |U3ytQVfgEmdIba03IlxzHTJL4NQn6WsZyzRzgLxJf2eJ2ACf7RX/Lf3tjW0OI+sq
+              |gQ7QpOVFPHficPfhiQkoThvHoc2kGMWL/WkENTPrcUZ6bhwIqJ+orK6dL4NGiJiB
+              |TuHOsJOv21y7l90fnaenL5lyBkMWeHfzYQkhGfcUN/55yWkjkrOkgeFUqXkyQQ1v
+              |Jq19A1ObBOe7axqsHvqeW4GJapXBjcWFRsq2ltP19lefKu27ikMWzEvkYJu1TX7t
+              |RmA82uwwqMa4eo5lXYkLCkVLyN3sSb39vqMb9C45y2fjBZv9OcpWCt/c0FwKz00K
+              |TZEW7+Ri43DsMyRwYiCafuVThL+J
+              |-----END CERTIFICATE-----""".stripMargin
+
+  val node1 = NodeConfigData.node1.modify(_.agentsName.each.securityToken).setTo(Certificate(cert1))
+  val node2 = NodeConfigData.node2.modify(_.agentsName.each.securityToken).setTo(Certificate(cert2))
+
+  val nodes = (NodeConfigData.root :: node1 :: node2 :: Nil).map(x => (x.id, x)).toMap
+
+  val dest = File("/tmp/rudder-test-allnodecerts.pem")
+
+  dest.deleteOnExit()
+
+  sequential // because files, etc
+
+  "When nodes have certificates, they" should {
+    "certificate are writen to file only for nodes with certificate" in {
+      val writer = new WriteNodeCertificatesPemImpl(Some(s"/usr/bin/touch ${dest.pathAsString}"))
+      val res = ZioRuntime.runNow(writer.writeCertificates(dest, nodes).either)
+
+      (res must beRight) and
+      (dest.contentAsString(StandardCharsets.UTF_8) must beEqualTo(cert1 + "\n" + cert2 + "\n"))
+    }
+  }
+
+  "when there is an error in async, we shoud get a log" in {
+
+    val writer = new WriteNodeCertificatesPemImpl(Some("/non/existing/command"))
+
+    import scala.collection.JavaConverters._
+
+    val log = writer.logger.logEffect._internal.asInstanceOf[ch.qos.logback.classic.Logger]
+
+    val ctx = LoggerFactory.getILoggerFactory.asInstanceOf[LoggerContext]
+    import ch.qos.logback.classic.encoder.PatternLayoutEncoder
+    val encoder = new PatternLayoutEncoder()  // this is mandatory. Without that, nothing happen silently
+    encoder.setContext(ctx)                   // order of setters matters
+    encoder.setPattern("%msg%n")
+    encoder.start()                           // don't forget start!
+
+    val os = new ByteArrayOutputStream()
+    val appender = new OutputStreamAppender[ILoggingEvent]()
+    appender.setImmediateFlush(true)
+    appender.setName("test-byte-array")
+    appender.setContext(ctx)                  // context must be set first
+    appender.setEncoder(encoder)              // then encoder
+    appender.setOutputStream(os)              // then output stream
+    appender.start()                          // don't forget start!
+
+    // if you want to remove other appenders
+    log.iteratorForAppenders().asScala.foreach { a =>
+      log.detachAppender(a)
+      a.stop()
+    }
+
+    log.setAdditive(false) // disable root logger/appender
+    log.addAppender(appender)
+
+    // exec
+    writer.writeCerticatesAsync(dest, nodes)
+
+    Thread.sleep(100)
+
+    val output = os.toString(StandardCharsets.UTF_8)
```

Comment:
toString(StandardCharsets) is not working in java8 ! 

See https://ci.normation.com/jenkins/job/test-rudder-pr/48/console